AssociativeCat.agda

```diff
@@ -0,0 +1,225 @@
+    there : ∀ {a as ss}
+          → SomeSolid {as} ss
+          → SomeSolid {a ∷ as} (Blank ∷ ss)
+
+  data Row : ∀ {as as′} → SolidityList as → SolidityList as′ → Set where
+    []    : ∀ {as ss}
+          → Row {as} ss (all Blank as)
+    blank : ∀ {a as as′ ss ss′}
+          → Row {as} {as′} ss ss′
+          → Row {a ∷ as} {a ∷ as′} (Blank ∷ ss) (Blank ∷ ss′)
+    piece : ∀ {t t′}
+          → C t t′
+          → let as₀  = ty→list t
+                as₀′ = ty→list t′
+         in {ss₀ : SolidityList as₀}
+          → SomeSolid ss₀
+          → let ss₀′ = all Solid as₀′
+         in ∀ {as as′ ss ss′}
+          → Row {as} {as′} ss ss′
+          → Row {as₀ ++ as} {as₀′ ++ as′} (ss₀ ++′ ss) (ss₀′ ++′ ss′)
+
+  data Pile : ∀ {as as′} → SolidityList as → SolidityList as′ → Set where
+    [] : ∀ {as ss}
+       → Pile {as} ss (all Blank as)
+    _∷_ : ∀ {as as′ as′′ ss ss′ ss′′}
+        → Row {as} {as′} ss ss′
+        → Pile {as′} {as′′} ss′ ss′′
+        → Pile {as} {as′′} ss ss′′
+
+  data FreeAssocCat (t t′ : Type) : Set where
+    MkFreeAssocCat : let as  = ty→list t
+                         as′ = ty→list t′
+                         ss  = all Solid as
+                         ss′ = all Blank as′
+                  in Pile ss ss′
+                   → FreeAssocCat t t′
+
+  module _ where
+    open ≈-Reasoning
+
+    nil : ∀ {t} → FreeAssocCat t t
+    nil = MkFreeAssocCat []
+    
+    postulate TODO : {X : Set} → X
+    
+    cons : ∀ {t t′ t′′}
+         → C t t′
+         → FreeAssocCat t′ t′′
+         → FreeAssocCat t t′′
+    cons = TODO  -- a generalization of (exp→free (embed f)) ?
+
+    ty-is-cons : ∀ t
+               → Σ[ a ∈ A ]
+                 Σ[ as ∈ List A ]
+                 ty→list t ≡ a ∷ as
+    ty-is-cons (atom a) = a , [] , refl
+    ty-is-cons (t₁ , t₂) = let a , as , prf = ty-is-cons t₁
+                            in a , as ++ ty→list t₂ , go a as prf
+      where
+        go : (a : A)
+           → (as : List A)
+           → (prf : ty→list t₁ ≡ a ∷ as)
+           → ty→list (t₁ , t₂) ≡ a ∷ (as ++ ty→list t₂)
+        go = TODO  -- doesn't look too hard...
+    
+    cons-is-solid : ∀ {a as} → SomeSolid (all Solid (a ∷ as))
+    cons-is-solid = here
+    
+    type-is-solid : ∀ t → SomeSolid (all Solid (ty→list t))
+    type-is-solid = TODO  -- use ty-is-cons to reduce the problem to cons-is-solid
+
+    exp→free : ∀ {t t′} → Exp t t′ → FreeAssocCat t t′
+    exp→free {t} {t′} (embed f) = MkFreeAssocCat (row′ ∷ [])
+      where
+        as₀ : List A
+        as₀ = ty→list t
+
+        as₀′ : List A
+        as₀′ = ty→list t′
+
+        ss₀ : SolidityList as₀
+        ss₀ = all Solid as₀
+
+        ss₀′ : SolidityList as₀′
+        ss₀′ = all Solid as₀′
+
+        row : Row {as₀ ++ []} {as₀′ ++ []} (ss₀ ++′ []) (ss₀′ ++′ [])
+        row = piece f (type-is-solid t) []
+        
+        row′ : Row {as₀} {as₀′} ss₀ ss₀′
+        row′ = TODO  -- prove that (++ []) is identity everywhere
```

Comment:
I thought SolidityList would make things easier, but I think it makes this part worse. Better stick with the `List (A , Solidity)` approach.